server/src/main/scala/io/delta/standalone/internal/DeltaSharingCDCReader.scala

```diff
@@ -0,0 +1,281 @@
+package io.delta.standalone.internal
+
+import java.sql.Timestamp
+
+import io.delta.standalone.DeltaLog
+import io.delta.standalone.internal.actions.{
+  AddCDCFile,
+  AddFile,
+  CommitInfo,
+  FileAction,
+  Metadata,
+  RemoveFile
+}
+import io.delta.standalone.internal.exception.DeltaErrors
+import io.delta.standalone.internal.util.ConversionUtils
+import org.apache.hadoop.conf.Configuration
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+/**
+ * A class to load Delta tables from `TableConfig`. It also caches the loaded tables internally
+ * to speed up the loading.
+ */
+class DeltaSharingCDCReader(val deltaLog: DeltaLogImpl, val conf: Configuration) {
+
+  private lazy val snapshot = deltaLog.snapshot
+  lazy val protocol = snapshot.protocolScala
+  lazy val metadata = snapshot.metadataScala
+  private lazy val history = DeltaHistoryManager(deltaLog)
+
+  private[internal] def getCDCVersions(
+      cdfOptions: Map[String, String],
+      latestVersion: Long): (Long, Long) = {
+    val startingVersion = getVersionForCDC(
+      cdfOptions,
+      DeltaDataSource.CDF_START_VERSION_KEY,
+      DeltaDataSource.CDF_START_TIMESTAMP_KEY,
+      latestVersion
+    )
+    if (startingVersion.isEmpty) {
+      throw DeltaCDFErrors.noStartVersionForCDF
+    }
+    // add a version check here that is cheap instead of after trying to list a large version
+    // that doesn't exist
+    if (startingVersion.get > latestVersion) {
+      throw DeltaCDFErrors.startVersionAfterLatestVersion(startingVersion.get, latestVersion)
+    }
+
+    val endingVersion = getVersionForCDC(
+      cdfOptions,
+      DeltaDataSource.CDF_END_VERSION_KEY,
+      DeltaDataSource.CDF_END_TIMESTAMP_KEY,
+      latestVersion
+    )
+
+    if (endingVersion.exists(_ < startingVersion.get)) {
+      throw DeltaCDFErrors.endBeforeStartVersionInCDF(startingVersion.get, endingVersion.get)
+    }
+
+    (startingVersion.get, endingVersion.getOrElse(latestVersion))
+  }
+
+  // Convert timestamp string in cdfOptions to Timestamp
+  private def getTimestamp(paramName: String, timeStampStr: String): Timestamp = {
+    try {
+      Timestamp.valueOf(timeStampStr)
+    } catch {
+      case e: IllegalArgumentException =>
+        throw DeltaCDFErrors.invalidTimestamp(paramName, e.getMessage)
+    }
+  }
+
+  /**
+   * - If a commit version exactly matches the provided timestamp, we return it.
+   * - Otherwise, we return the earliest commit version
+   *   with a timestamp greater than the provided one.
+   * - If the provided timestamp is larger than the timestamp
+   *   of any committed version, we throw an error.
+   *
+   * @param timestamp - user specified timestamp
+   * @param latestVersion - latest version of the table
+   * @return - corresponding version number for timestamp
+   */
+  private def getStartingVersionFromTimestamp(timestamp: Timestamp, latestVersion: Long): Long = {
```

Comment:
Could this be refactored into DeltaHistoryManager itself?

Reply:
Technically it could, though I prefer not. What's the main benefit?

The main reason to not refactor it into DeltaHistoryManager is to keep consistency with the code structure in the databricks server and DBR, where  `getStartingVersionFromTimestamp` is not a function of `DeltaHistoryManager`.

Reply:
The reasons is I suggest is:
   1) Looking at the code, it does not need any other state. 
   2) All version manipulations are in the same history object.
   3) We can refactor.

We could potentially refactor it in dbr as well (perhaps in a separate PR).

Reply:
I see, it makes sense. Just realized it's `DeltaHistoryManager` not `DeltaSharingHistoryManager`, which is under delta-io/connectors instead of delta-io/delta-sharing, which needs a separate release after modification.

Reply:
Ok, then perhaps let's add a todo for now.

Reply:
sg https://github.com/delta-io/delta-sharing/issues/144